File: downloads/code/LeetCode/Java/Sort-List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode sortList(ListNode head) {
        if(head==null || head.next==null) return head;
        // 找到中间点
        ListNode slow = head;
        ListNode fast = head;
        ListNode prev = null;
        while(fast!=null && fast.next!=null){
            prev = slow;
            slow = slow.next;
            fast = fast.next.next;
        }

        prev.next = null;

        // Divide
        ListNode left = sortList(head);
        ListNode right = sortList(slow);

        // Merge
        return merge(left,right);
    }

    ListNode merge(ListNode l1, ListNode l2){
        ListNode ret = new ListNode(0);
        ListNode cur = ret;
        while(l1!=null && l2!=null){
            if(l1.val < l2.val){
                cur.next = l1;
                l1 = l1.next;
            }
            else {
                cur.next = l2;
                l2 = l2.next;
            }
            cur = cur.next;
        }
        if (l1!=null)
            cur.next = l1;

        if (l2!=null)
            cur.next = l2;
        
        return ret.next;
    }

}

File: downloads/code/LeetCode/Python/Sort-List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def sortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head == None or head.next==None:
            return head

        # merge sort
        fast_node = head
        slow_node = head

        # find pivot
        while fast_node.next != None and fast_node.next.next != None:
            slow_node = slow_node.next
            fast_node = fast_node.next.next
        
        middle = slow_node.next
        slow_node.next = None
        
        # split
        leftNode = self.sortList(head)
        rightNode = self.sortList(middle)

        return self.merge(leftNode,rightNode)

    # merge
    def merge(self,left,right):
        if left != None and right == None:
            return left
        if left == None and right != None:
            return right
        if left == None and righ == None:
            return None
        
        # find head 
        ret = None
        if left.val < right.val:
            ret = left
            left = left.next
        else:
            ret = right
            right = right.next
        
        # merge every node
        head = ret
        while left != None and right != None:
            if left.val < right.val:
                head.next = left
                left = left.next
            else:
                head.next = right
                right = right.next
            head = head.next
        
        if left != None:
            head.next = left
        if right != None:
            head.next = right
       
        return ret

